### PR TITLE
PIM-8673: Add a fallback to get the mime-type of files loaded without metadata.

### PR DESCRIPTION
**Pull request by pchasle**
**Description (for Contributor and Core Developer)**

When using a cloud storage solution (AWS S3, Google Cloud Storage). You need to send the `Content-Type` metadata along with your file or it will be stored as `application/octet-stream`.

The goal of this PR is to:
- Set the metadata option required by each Flysystem adapter (dirty fix, but a rewrite to handle properly each adapter configuration would probably be too much).
- Fallback to load the mime-type of a file from the FileInfo repository if we get a `Content-Type` `application/octet-stream` when loading a file from the storage.

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Todo
| Added legacy Behats               | Todo
| Added acceptance tests            | Todo
| Added integration tests           | Todo
| Changelog updated                 | Todo
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Akeneo/Platform/Bundle/UIBundle/Imagine/FlysystemLoader.php [hunk 1]
```diff
@@ -85,6 +86,13 @@ protected function retrieveContentFileFromVfs($path)
             throw new NotLoadableException(sprintf('Unable to read the file "%s" from the filesystem.', $path));
         }
 
+        if ('application/octet-stream' === $mimeType && null !== $this->fileInfoRepository) {
```
Comment (ahocquard) [inline src/Akeneo/Platform/Bundle/UIBundle/Imagine/FlysystemLoader.php]: not tested (a specification exist)

File: src/Akeneo/Platform/Bundle/UIBundle/Imagine/FlysystemLoader.php [hunk 2]
```diff
@@ -23,14 +24,14 @@ class FlysystemLoader implements LoaderInterface
...
 
-    /**
-     * @param FilesystemProvider $filesystemProvider
-     * @param array              $filesystemAliases
-     */
-    public function __construct(FilesystemProvider $filesystemProvider, array $filesystemAliases)
+    /** @var FileInfoRepositoryInterface | null */
+    protected $fileInfoRepository;
+
+    public function __construct(FilesystemProvider $filesystemProvider, array $filesystemAliases, FileInfoRepositoryInterface $fileInfoRepository = null)
```
Comment (ahocquard) [inline src/Akeneo/Platform/Bundle/UIBundle/Imagine/FlysystemLoader.php]: could you add a comment  to remove null during merge please

Comment (mmetayer) [inline src/Akeneo/Platform/Bundle/UIBundle/Imagine/FlysystemLoader.php]: please add a TODO to remove the `= null` for the last argument on 4.0

File: src/Akeneo/Platform/Bundle/UIBundle/Imagine/FlysystemLoader.php (hunk 1 above)
Comment (mmetayer) [inline src/Akeneo/Platform/Bundle/UIBundle/Imagine/FlysystemLoader.php]: Could you add a test? Also, add a TODO to remove the ` null !== $this->fileInfoRepository` test on 4.0

File: src/Akeneo/Platform/Bundle/UIBundle/Imagine/FlysystemLoader.php (hunk 2 above)
Comment (pchasle) [reply src/Akeneo/Platform/Bundle/UIBundle/Imagine/FlysystemLoader.php]: This fallback could stay optional as it's mainly a patch to fix a wrong mime-type on existing objects. What do you think?

Comment (ahocquard) [reply src/Akeneo/Platform/Bundle/UIBundle/Imagine/FlysystemLoader.php]: if it's for a patch and useless later on: the argument should be removed on master
if it's mandatory to do it later on: remove null

But definitively not an optional dependency on master :)

Comment (mmetayer) [reply src/Akeneo/Platform/Bundle/UIBundle/Imagine/FlysystemLoader.php]: I don't like optional arguments in constructors, it forces us to test the existence of the property before being able to use it (see L89: `&& null !== $this->fileInfoRepository`)

Comment (pchasle) [reply src/Akeneo/Platform/Bundle/UIBundle/Imagine/FlysystemLoader.php]: Thanks, I will add the comments to make the dependency mandatory on master.
